hs-bindgen/src-internal/HsBindgen/C/Parser.hs

```diff
@@ -130,6 +133,39 @@ parseCHeaders tracer args predicate _extSpec headerIncludePaths =
       headerIncludePath <- headerIncludePaths !? (singleLocLine sloc - 1)
       return $ ParseCHeadersInputFileNotFound headerIncludePath
 
+    -- This function resolves main header paths before we parse them.  There
+    -- should be little overhead since the same translation unit is used.
+    --
+    -- This is necessary to implement the @SelectFromMainFiles@ predicate.  If
+    -- a user specifies headers @a.h@ and @b.h@, where @a.h@ includes @b.h@,
+    -- then we need to know that @b.h@ is a main header before the inclusion
+    -- directive in the root header is processed.
+    --
+    -- The 'CHeaderIncludePath' is retrieved from the list via line number
+    -- because we are unable to distinguish between system and quote inclusions
+    -- from the AST (without re-parsing the macro).
+    resolveMainHeaders ::
+         CXCursor
+      -> IO (Next IO (SourcePath, CHeaderIncludePath))
+    resolveMainHeaders cursor = either return return <=< runExceptT $ do
+      loc <- clang_getCursorLocation cursor
+      sloc <- HighLevel.clang_getExpansionLocation loc
+      -- skip builtin macros
+      when (nullSourcePath (singleLocPath sloc)) $ throwError (Continue Nothing)
+      -- only process the root header
+      isFromRootHeader <- clang_Location_isFromMainFile loc
+      unless isFromRootHeader $ throwError (Break Nothing)
```

Comment:
This is wrong: `#include` directives don't have children, so this would cause the traversal to miss includes that might come later.